lib/Sema/TypeCheckDecl.cpp

```diff
@@ -5581,6 +5583,152 @@ static Optional<ObjCReason> shouldMarkClassAsObjC(TypeChecker &TC,
   return None;
 }
 
+bool checkDeclCircularityImpl(Decl *decl, std::set<TypeBase *> &known);
+
+// Given a type, check if its in 'known'. Do the same with its subtypes.
+// When the type is nominal, get its declaration and check if any if its
+// member's type is in known by calling 'checkDeclCircularityImpl'
+bool isTypePresentForCircularity(Type type, std::set<TypeBase *> &known) {
+  auto desugared = type->getDesugaredType();
+
+  if (auto NT = dyn_cast<NominalType>(desugared)) {
+    if(auto NTT = NT->getDecl()->getDeclaredTypeInContext()) {
+      if (known.find(NTT.getPointer()) != known.end()) {
+        return true;
+      }
+    }
+    // the type is cleared, now we check its member.
+    return checkDeclCircularityImpl(NT->getDecl(), known);
+  }
+
+  // check each generic argument, this takes care of parameter from declaring
+  // context as well as stuff like Optionals
+  if (auto BGT = dyn_cast<BoundGenericType>(desugared)) {
+    for (auto T: BGT->getGenericArgs()) {
+      if(isTypePresentForCircularity(T, known)) {
+        return true;
+      }
+    }
+
+    // check the overall generic type
+    auto BGTT = BGT->getDecl()->getDeclaredTypeInContext();
+    if (known.find(BGTT.getPointer()) != known.end()) {
+      return true;
+    }
+
+    // check this generic type's members.
+    return checkDeclCircularityImpl(BGT->getDecl(), known);
+  }
+
+  // if this is a tuple, deconstruct it and check each element
+  if (auto tuple = dyn_cast<TupleType>(desugared)) {
+    for (auto Elt: tuple->getElements()) {
+      if (isTypePresentForCircularity(Elt.getType(), known)) {
+        return true;
+      }
+    }
+  }
+
+  if (known.find(desugared) != known.end()) {
+    return true;
+  }
+  return false;
+}
+
+
+// Given a declaration, iterate over its members and make sure non of them
+// is of a value type of the very same declaration. Use
+// 'isTypePresentForCircularity' to exam related types of each member.
+bool checkDeclCircularityImpl(Decl *decl, std::set<TypeBase *> &known) {
+
+  // skip unintesting declarations...
+
+  // ...implicit members...
+  if (decl->isImplicit()) { return false; }
+
+  // ...static memebers, non-value types...
+  if (auto s = dyn_cast<VarDecl>(decl)) {
+    if (s->isStatic() || !s->hasStorage()) { return false; }
+  }
+
+  bool isEnum = false;
+  bool isStruct = isa<StructDecl>(decl);
+
+  // ... and enums marked 'indirect'
+  if (auto e = dyn_cast<EnumDecl>(decl)) {
+    if (e->isIndirect()) {
+      return false;
+    }
+    isEnum = true;
+  }
+
+  if (auto nominal = dyn_cast<NominalTypeDecl>(decl)) {
+    auto dt = nominal->getDeclaredTypeInContext();
+    // record existence of this struct or enum, if we see it in value member
+    // types, subtypes, etc, complain.
+    if (isEnum || isStruct) {
+      known.insert(dt.getPointer());
+    }
+
+    // recurse for nested struct/enums as well as normal member declarations.
+    for (auto m: nominal->getMembers(false)) {
+      if (isa<VarDecl>(m) ||
```

Comment:
Is the 'if' really necessary? You already return 'false' if the decl isn't one of those types inside checkDeclCircularityImpl()